src/foundry/client-esm/dice/roll.d.mts

```diff
@@ -491,20 +490,24 @@ declare class Roll<D extends Record<string, unknown> = {}> {
    * @returns A promise which resolves to the created ChatMessage entity, if create is true
    *          or the Object of prepared chatData otherwise.
    */
-  toMessage<T extends DeepPartial<ConstructorParameters<ConfiguredDocumentClass<typeof ChatMessage>>[0]> = EmptyObject>(
+  toMessage<T extends foundry.documents.BaseChatMessage.ConstructorData = EmptyObject, Create extends boolean = true>(
     messageData?: T,
-    { rollMode, create }?: InexactPartial<{ rollMode: keyof CONFIG.Dice.RollModes | "roll"; create: true }>,
-  ): Promise<ChatMessage.ConfiguredInstance | undefined>;
-
-  toMessage<T extends DeepPartial<ConstructorParameters<ConfiguredDocumentClass<typeof ChatMessage>>[0]> = EmptyObject>(
-    messageData: T,
-    { rollMode, create }: { rollMode?: keyof CONFIG.Dice.RollModes | "roll" | undefined; create: false },
-  ): Roll.MessageData<T>;
-
-  toMessage<T extends DeepPartial<ConstructorParameters<ConfiguredDocumentClass<typeof ChatMessage>>[0]> = EmptyObject>(
-    messageData: T,
-    { rollMode, create }: { rollMode?: keyof CONFIG.Dice.RollModes | "roll" | undefined; create: boolean },
-  ): Promise<ChatMessage.ConfiguredInstance | undefined> | Roll.MessageData<T>;
+    options?: InexactPartial<{
+      /**
+       * The template roll mode to use for the message from CONFIG.Dice.rollModes
+       * @remarks "roll" equivalent to explicit undefined
+       */
+      rollMode: keyof CONFIG.Dice.RollModes | "roll";
+      /**
+       * Whether to automatically create the chat message, or only return the prepared chatData object.
+       * @defaultValue `true`
+       */
+      create: Create;
+    }>,
+  ): Promise<
+    | (true extends Create ? ChatMessage.ConfiguredInstance | undefined : never)
+    | (false extends Create ? Roll.MessageData<T> : never)
```

Comment:
Unfortunately in the case of `true | false` (i.e. `boolean`) this means you'd get `Promise<never>`. So `Create extends true`/`Create extends false` is probably better.

Reply:
Both seem to work for this test (which is already in the PR)
```ts
expectTypeOf(r.toMessage({}, { create: testBool })).toEqualTypeOf<
  // eslint-disable-next-line @typescript-eslint/no-empty-object-type
  Promise<ChatMessage | undefined | foundry.dice.Roll.MessageData<{}>>
>();
```